README.md
# cinnamon-plugin-asl-protocol
A Cinnamon plugin that adds convenience methods for Apollo Software's REST
req/res protocol.

## Installation
1. `yarn add @apollosoftwarexyz/cinnamon-plugin-asl-protocol`
2. Update your Cinnamon project's `src/main.ts` to include the plugin in the
   `load` hook:
  ```ts
  import { ApolloProtocol } from '@apollosoftwarexyz/cinnamon-plugin-asl-protocol';
  
  // ...
  
  await Cinnamon.initialize({
      async load(framework) {
          // Adds convenience methods for common response types.
          framework.use(new ApolloProtocol(framework));
      }
  });
  
  // ...
  ```
  
## Usage
You can use the `success`, `successRaw` and `error` methods in your controller
routes, by calling the respective method on the context.

- For a `success` response:
  
  ```ts
  @Route(Method.GET, '/')
  public async index(ctx: Context): Promise<void> {
  
    // ...
    
    return ctx.success({
        message: 'Hello, world!'
    });
  
  }
  ```
  
  This will yield the following JSON response, which would usually be unwrapped
  on the client-side by the complimentary client plugin:
  ```json
  {
    "success": true,
    "payload": {
      "message": "Hello, world!"
    }
  }
  ```

- For a `successRaw` response:
  
  ```ts
  @Route(Method.GET, '/')
  public async index(ctx: Context): Promise<void> {
  
    // ...
    
    // The MIME-type 'text/plain' would be set by default, but is specified
    // here for demonstration purposes.
    return ctx.successRaw("Hello, world!", 'text/plain');
  
  }
  ```
  
  This will yield the following plain-text response:
  ```
  Hello, world!
  ```

- For an `error` response:
  
  ```ts
  @Route(Method.GET, '/')
  public async index(ctx: Context): Promise<void> {
  
    // ...
    
    // Where:
    // - 404 is the HTTP status code,
    // - "ERR_MISSING_ENTITY" is your app-specific error name
    // - "User not found!" is a human-readable message in your product's
    //   primary language that your app can display as a fallback to its local
    //   translation based on the error name.
    return ctx.error(404, "ERR_MISSING_ENTITY", "User not found!");
  
  }
  ```
  
  This will yield the following JSON response:
  ```json
  {
    "success": false,
    "error": "ERR_MISSING_ENTITY",
    "message": "User not found!"
  }
  ```

## License
[MIT License](./LICENSE)


CHANGELOG.md
# v2.0.3
- Replace deprecated `CinnamonWebServerModulePlugin` with
  `CinnamonHookConsumer`
- Replace `_Context` (removed) with `BaseContext`
- Cinnamon v0.2.0+ only 

# v2.0.2
- Fix typing issues
- Update for Cinnamon v0.2.0
- Bump dependencies

# v2.0.1
- Bump dependencies

# v2.0.0
- Update for Cinnamon v0.1.0 (alpha->beta)
- Remove Koa as a dependency
- Version 2 of the Apollo Software Limited Protocol
  - Add HTTP header for protocol version
  - Add automatic error and response handling
- Remove `body` from `Context` and `Response` (to encourage use of the
  `success`, `successRaw` and `error` methods)
- Add optional `statusCode` (HTTP status code) parameter to `success` and
  `successRaw`.
- Bump dependencies

# v1.0.4
- Add yarn versioning plugin

# v1.0.3
- Update for Cinnamon v0.0.36 (final alpha build)
- Bump dependencies

# v1.0.2
- Changes to metadata of package

# v1.0.1
- Update for Cinnamon v0.0.21

# v1.0.0
- Initial release


types/shim.d.ts
export * from './main';
import { BaseContext } from '@apollosoftwarexyz/cinnamon';

type Response = Omit<BaseContext["response"], "body">;

declare module "@apollosoftwarexyz/cinnamon" {
    interface Context {
        /**
         * You should not set the response body directly with
         * cinnamon-plugin-asl-protocol.
         *
         * You should instead use ctx.successRaw, or access the response stream
         * directly with ctx.res.
         */
        body: never;

        /**
         * You should not set the response body directly with
         * cinnamon-plugin-asl-protocol.
         *
         * You should instead use ctx.successRaw, or access the response stream
         * directly with ctx.res.
         */
        response: Response & {
            /**
             * You should not set the response body directly with
             * cinnamon-plugin-asl-protocol.
             *
             * You should instead use ctx.successRaw, or access the response stream
             * directly with ctx.res.
             */
            body: never;
        };

        /**
         * Indicates the response was successful by JSON-encoding and wrapping
         * the specified payload with an ASL Protocol Success response before
         * returning it to the client.
         *
         * @param payload Optionally, the response payload to include.
         * @param statusCode Optionally, the status code to set. (Default: 200)
         */
        success: (payload?: any, statusCode?: number) => void;

        /**
         * Indicates that the request was successful by returning the raw
         * payload to the client by inserting `raw=true` into your payload if
         * its an object or directly returning the payload to the client if it
         * isn't.
         *
         * @param rawPayload Optionally, the raw payload to respond. (`raw=true`
         *                   will be injected if it is an object.)
         * @param mimeType The MIME type to set in the response header. Defaults
         *                 to `text/plain`.
         * @param statsuCode Optionally, the status code to set. (Default: 200)
         */
        successRaw: (rawPayload?: any, mimeType?: string, statusCode?: number) => void;

        /**
         * Indicates that the request failed by returning an ASL Protocol Error
         * response payload with the specified parameters to the client.
         *
         * @param number The HTTP status code most relevant to the error.
         *               Defaults to `400`.
         * @param error The name of the error that occurred. Defaults to
         *              `ERR_UNEXPECTED`.
         * @param message A human-readable message explaining why
         *                the error occurred. Defaults to 'An unexpected error
         *                occurred.'
         */
        error: (code?: number, error?: string, message?: string) => void;
    }
}



src/main.ts
import Cinnamon, {
    HttpError,
    CinnamonPlugin,
    CinnamonHookConsumer,
    WebServer,
} from '@apollosoftwarexyz/cinnamon';

export const ASW_PROTOCOL_VERSION = 2;

export class ApolloProtocol extends CinnamonPlugin implements CinnamonHookConsumer {
    constructor(framework: Cinnamon, options?: {}) {
        super(framework, "xyz.apollosoftware", "cinnamon.protocol");
    }

    async onInitialize() {
        return true;
    }

    async beforeRegisterControllers() {
        this.framework
            .getModule<WebServer>(WebServer.prototype)
            .server.use(async (ctx, next) => {
            ctx.success = function (payload?: any, statusCode: number = 200) {
                (ctx as any).__hasApolloResponse = true;
                ctx.set(
                    "x-apollo-software-protocol",
                    ASW_PROTOCOL_VERSION.toString()
                );

                ctx.response.status = statusCode;
                ctx.response.type = "application/json";
                ctx.response.body = JSON.stringify({
                    success: true,
                    payload,
                });
            };

            ctx.successRaw = function (
                rawPayload: any,
                mimeType: string = "text/plain",
                statusCode: number = 200
            ) {
                (ctx as any).__hasApolloResponse = true;

                ctx.response.status = statusCode;
                ctx.response.type = mimeType;
                ctx.response.body = rawPayload;
            };

            ctx.error = function (
                code: number = 400,
                error: string = "ERR_UNEXPECTED",
                message: string = "An unexpected error occurred."
            ) {
                (ctx as any).__hasApolloResponse = true;
                ctx.set(
                    "x-apollo-software-protocol",
                    ASW_PROTOCOL_VERSION.toString()
                );

                ctx.response.status = code;
                ctx.response.type = "application/json";
                ctx.response.body = JSON.stringify({
                    success: false,
                    error,
                    message,
                });
            };

            // Call the request handlers defined after this one.
            await next();

            // If the request has already been handled by some handler from
            // this plugin, we'll just return here and do nothing.
            if ((ctx as any).__hasApolloResponse) return;

            // Otherwise...

            // Add support for Cinnamon's 'return-response' functionality,
            // by ensuring that if, after the request handlers were called,
            // either an ASL Protocol Response was returned or the response
            // was converted to one that conformed anyway.

            // If the request was a 404, we'll ensure that JSON is returned
            // regardless of whether it was requested (provided, of course,
            // an explicit handler hasn't been provided).
            if (ctx.status === 404) {
                return ctx.error(
                    404,
                    'ERR_NOT_FOUND',
                    "The resource you requested could not be found."
                );
            }

            // If a successful response is set and marked as JSON,
            // simply wrap it (if it needs wrapping).
            if (ctx.status < 400) {

                // If there is a body present, either ensure it is
                // already in a valid form, or attempt to convert it.
                if (ctx.body) {

                    if (ctx.response.is('application/json')) {
                        // If the response object does not have a
                        // success property on it, wrap it with a
                        // success response and return it.
                        if (!Object.hasOwn(ctx.body, "success")) {
                            // If there is a payload object, we'll
                            // return it as though it were a top-level
                            // object.
                            //
                            // (If the developer truly wanted to return
                            // the top level object, it's expected that
                            // they would set success too.)
                            if (Object.hasOwn(ctx.body, "payload")) {
                                return ctx.success(
                                    ctx.body.payload,
                                    ctx.status
                                );
                            }

                            // Otherwise, we'll just assume that the
                            // return value was the payload itself.
                            return ctx.success(ctx.body, ctx.status);
                        }

                        // If the success property is not true, return
                        // an error response.
                        if (!ctx.body.success) {
                            // There's no real way to determine why the
                            // response wasn't successful, or if it
                            // indeed wasn't, so we'll return a vauge
                            // error here for the developer's benefit.
                            return ctx.error(
                                500,
                                'ERR_INVALID_RESPONSE',
                                'The server produced an invalid response.'
                            );
                        }

                        // Otherwise, either wrap the object if has no
                        // payload property, otherwise return as-is.
                        if (!Object.hasOwn(ctx.body, "payload")) {
                            // Remove the success property, it'll get
                            // re-added at the top level.
                            delete ctx.body.success;

                            // Then, return a success response with
                            // the returned object included as the
                            // 'payload' value.
                            return ctx.success(ctx.body, ctx.status);
                        } else return;
                    }

                        // If it's not a JSON body, we can simply return
                    // the body as a successRaw response.
                    else {
                        return ctx.successRaw(
                            ctx.body,
                            ctx.type,
                            ctx.status
                        );
                    }

                }

                // Otherwise, simply return an empty success response.
                return ctx.success(undefined, ctx.status);

            }

            // Otherwise, if there's an error object, or the response
            // has a code above 400 and below 600 AND the framework is
            // in production mode...
            else if (
                (ctx.errorObject || (ctx.status >= 400 && ctx.status < 600)) &&
                !this.framework.inDevMode
            ) {
                // If the response body is JSON, defines an error
                // object and has success set to false, simply do
                // nothing.
                if (ctx.request.is('application/json')) {
                    if (
                        Object.hasOwn(ctx.body, "success") &&
                        Object.hasOwn(ctx.body, "error")
                    ) return;
                }

                // ...if not, show the errorObject if it's present...
                if (
                    ctx.errorObject &&
                    ctx.errorObject instanceof HttpError
                ) {
                    return ctx.error(
                        ctx.errorObject.status,
                        "ERR_UNEXPECTED",
                        ctx.errorObject.message
                    );
                }

                // ...otherwise show a generic error.
                ctx.error(
                    ctx.status,
                    "ERR_UNEXPECTED",
                    "An unexpected error has occurred."
                );
            }
        });
    }
}
